lib/auth/grpcserver_test.go

```diff
@@ -2532,3 +2534,103 @@ func TestExport(t *testing.T) {
 		})
 	}
 }
+
+// TestSAMLValidation tests that SAML validation does not perform an HTTP
+// request if the calling user does not have permissions to create or update
+// a SAML connector.
+func TestSAMLValidation(t *testing.T) {
+	modules.SetTestModules(t, &modules.TestModules{
+		TestFeatures: modules.Features{SAML: true},
+	})
+
+	// minimal entity_descriptor to pass validation. not actually valid
+	const minimalEntityDescriptor = `
+<md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" entityID="http://example.com">
+  <md:IDPSSODescriptor>
+    <md:SingleSignOnService Location="http://example.com" />
+  </md:IDPSSODescriptor>
+</md:EntityDescriptor>`
+
+	allowSAMLUpsert := types.RoleConditions{
+		Rules: []types.Rule{{
+			Resources: []string{types.KindSAML},
+			Verbs:     []string{types.VerbCreate, types.VerbUpdate},
+		}},
+	}
+
+	testCases := []struct {
+		desc             string
+		allow            types.RoleConditions
+		entityDescriptor string
+		edURLCalled      bool
+		errAs            error
```

Comment:
Maybe:

```suggestion
		entityDescriptor string
		assertErr        func(error) bool
```

So you can pass in `trace.IsAccessDenied` or `trace.IsBadParameter` directly? (Just be careful with `nil` in the test body.)

Reply:
done. no special nil handling required, except for what was already there. `trace.IsBadParameter(nil)` is valid code. Added an explicit `require.Error(t, err)` so the error message is clear if we didn't get an error at all.